### PR TITLE
Add Sync_SSO

### LINKED ISSUE #483: Add sync_sso
Add sync_sso.
sync_sso is a url endpoint in Discourse (forum) to update user accounts (username, email...).
We need this feature so if name changes get added in the feature they are automatically handled by Ore.

[More info](https://meta.discourse.org/t/official-single-sign-on-for-discourse-sso/13045#heading--syncronize_records)

### PR DESCRIPTION
**Pull request by Ewoutvans**
Add sync_sso

added 
- `/api/sync_sso`
- added `apikey` to config
- added `secret` to config

Closes #483

### CODE REVIEW

File: app/controllers/ApiController.scala [hunk 1]
```diff
@@ -258,4 +261,64 @@ final class ApiController @Inject()(api: OreRestfulApi,
...
+
+        val confApiKey = this.config.security.get[String]("sync_sso.apikey")
+        val confSecret = this.config.security.get[String]("sync_sso.secret")
+
+        if (apiKey != confApiKey) {
+          BadRequest("API Key not valid")
+        } else if (!CryptoUtils.hmac_sha256(confSecret, sso.getBytes("UTF-8")).equals(sig)) {
+          BadRequest("Signature not matched")
+        } else {
+          val decoded = URLDecoder.decode(new String(Base64.getMimeDecoder.decode(sso)), "UTF-8")
```
Comment (Katrix) [inline app/controllers/ApiController.scala]: Play (through Akka) has nicer ways of dealing with query strings. Without looking too much into it, something like `Uri.Query(Base64.getMimeDecoder.decode(sso))` (charset is UTF-8 by default) is what you want to go with. Then call toMap on the result of that. From there, no more vars. Also as another tip, Scala provides a `toInt` method on strings to parse them as ints.

File: conf/application.conf.template [hunk 2]
```diff
@@ -66,6 +66,13 @@ security {
         secret = ${?SPONGE_AUTH_SSO_SECRET}
         timeout = 10000ms
     }
+
+    sync_sso {
+        secret = "changeme"
+        secret = ${?SPONGE_AUTH_SSO_SYNC_SECRET}
```
Comment (progwml6) [inline conf/application.conf.template]: we only need one SSO secret, remove this one

File: conf/application.conf.template [hunk 3]
```diff
@@ -66,6 +66,13 @@ security {
...
+        apikey = "changeme"
+        apikey = ${?SPONGE_AUTH_SSO_SYNC_APIKEY}
```
Comment (progwml6) [inline conf/application.conf.template]: call this SPONGE_AUTH_SSO_API_KEY and move it up to the sso object

**Review by progwml6 — CHANGES_REQUESTED** [review]

File: app/controllers/ApiController.scala [hunk 4]
```diff
@@ -258,4 +262,48 @@ final class ApiController @Inject()(api: OreRestfulApi,
...
+
+          val external_id = query.get("external_id")
+          val email = query.get("email")
+          val username = query.get("username")
+          val name = query.get("name")
+          val avatar_url = query.get("avatar_url")
+
+          val optUser = this.users.get(external_id.getOrElse("-1").toInt)
+
+          if (optUser.isEmpty) {
```
Comment (Katrix) [inline app/controllers/ApiController.scala]: This statement does nothing in this position. Also not a fan of testing the content in this way. I'd say the best here would be a pattern match on optUser.

File: app/controllers/ApiController.scala [hunk 5]
```diff
@@ -258,4 +262,48 @@ final class ApiController @Inject()(api: OreRestfulApi,
...
+          val external_id = query.get("external_id")
+          val email = query.get("email")
+          val username = query.get("username")
+          val name = query.get("name")
+          val avatar_url = query.get("avatar_url")
+
+          val optUser = this.users.get(external_id.get.toInt)
+
+          if (optUser.isEmpty) {
+            BadRequest(Json.obj("errors" -> Seq("User not found")))
```
Comment (Katrix) [inline app/controllers/ApiController.scala]: This statement does not do anything in this position

File: app/controllers/ApiController.scala [hunk 6]
```diff
@@ -258,4 +262,48 @@ final class ApiController @Inject()(api: OreRestfulApi,
...
+          }
+
+          val user = optUser.get
+
+          email.map(user.email_=)
```
Comment (Katrix) [inline app/controllers/ApiController.scala]: Prefer foreach for these instead of map

File: app/controllers/ApiController.scala [hunk 7]
```diff
@@ -258,4 +262,48 @@ final class ApiController @Inject()(api: OreRestfulApi,
...
+        val confSecret = this.config.security.get[String]("sso.secret")
+
+        if (apiKey != confApiKey) {
+          BadRequest("API Key not valid")
+        } else if (!CryptoUtils.hmac_sha256(confSecret, sso.getBytes("UTF-8")).equals(sig)) {
+          BadRequest("Signature not matched")
+        } else {
+          val query = Uri.Query(Base64.getMimeDecoder.decode(sso))
+
+          val external_id = query.get("external_id")
```
Comment (Katrix) [inline app/controllers/ApiController.scala]: Bit unsure if we instead want to use normal camel case for these values

File: app/controllers/ApiController.scala [hunk 8]
```diff
@@ -258,4 +262,48 @@ final class ApiController @Inject()(api: OreRestfulApi,
...
+        val sig = formData._2
+        val apiKey = formData._3
+
+
+        val confApiKey = this.config.security.get[String]("sso.apikey")
+        val confSecret = this.config.security.get[String]("sso.secret")
+
+        if (apiKey != confApiKey) {
+          BadRequest("API Key not valid")
+        } else if (!CryptoUtils.hmac_sha256(confSecret, sso.getBytes("UTF-8")).equals(sig)) {
```
Comment (Katrix) [inline app/controllers/ApiController.scala]: Use `==` instead of equals

File: app/controllers/ApiController.scala [hunk 9]
```diff
@@ -1,7 +1,9 @@
 package controllers
 
-import java.util.UUID
+import java.net.URLDecoder
```
Comment (Katrix) [inline app/controllers/ApiController.scala]: Check unused imports

**Review by Katrix — CHANGES_REQUESTED** [review]

**Pull request merged**